test/mail-test-srv/main.go

```diff
@@ -0,0 +1,225 @@
+	var toAddr []string
+	var msgBuf bytes.Buffer
+
+	clearState := func() {
+		fromAddr = ""
+		toAddr = nil
+		msgBuf = bytes.Buffer{}
+	}
+
+	var quit = false
+	for !quit {
+		line, err := getLine(readBuf)
+		if err != nil {
+			log.Printf("mailcatch: %s: getline: %v\n", conn.RemoteAddr(), err)
+			return
+		}
+
+		cmdSplit := strings.SplitN(line, " ", 2)
+		cmd := cmdSplit[0]
+		switch cmd {
+		case "QUIT":
+			quit = true
+			conn.Write([]byte("221 Bye \r\n"))
+			break
+		case "RSET":
+			clearState()
+			conn.Write(smtpOk250)
+		case "NOOP":
+			conn.Write(smtpOk250)
+		case "MAIL":
+			clearState()
+			matches := mailFromRegex.FindStringSubmatch(line)
+			if matches == nil {
+				log.Printf("mailcatch: %s: MAIL FROM parse error\n", conn.RemoteAddr())
+				conn.Write(smtpErr501)
+				continue
+			}
+			addr, err := mail.ParseAddress(matches[1])
+			if err != nil {
+				log.Printf("mailcatch: %s: addr parse error: %v\n", conn.RemoteAddr(), err)
+				conn.Write(smtpErr501)
+				continue
+			}
+			fromAddr = addr.Address
+			conn.Write(smtpOk250)
+		case "RCPT":
+			matches := rcptToRegex.FindStringSubmatch(line)
+			if matches == nil {
+				conn.Write(smtpErr501)
+				continue
+			}
+			addr, err := mail.ParseAddress(matches[1])
+			if err != nil {
+				log.Printf("mailcatch: %s: addr parse error: %v\n", conn.RemoteAddr(), err)
+				conn.Write(smtpErr501)
+				continue
+			}
+			toAddr = append(toAddr, addr.Address)
+			conn.Write(smtpOk250)
+		case "DATA":
+			conn.Write([]byte("354 Start mail input \r\n"))
+
+			for {
+				line, err := getLine(readBuf)
+				if err != nil {
+					log.Printf("mailcatch: %s: getline: %v\n", conn.RemoteAddr(), err)
+					return
+				}
+				msgBuf.WriteString(line)
+				msgBuf.WriteString("\r\n")
+				if strings.HasSuffix(msgBuf.String(), "\r\n.\r\n") {
+					break
+				}
+			}
+
+			mailResult := rcvdMail{
+				From: fromAddr,
+				Mail: msgBuf.String(),
+			}
+			allMailMutex.Lock()
+			for _, rcpt := range toAddr {
+				mailResult.To = rcpt
+				allReceivedMail = append(allReceivedMail, mailResult)
+				log.Printf("mailcatch: Got mail: %s -> %s\n", fromAddr, rcpt)
+			}
+			allMailMutex.Unlock()
+			conn.Write([]byte("250 Got mail \r\n"))
+			clearState()
+		}
+	}
+}
+
+func serveSMTP(l net.Listener) error {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			return err
+		}
+		go handleConn(conn)
+	}
+}
+
+func calculateAuth(c cmd.Config) error {
+	var buf bytes.Buffer
+	enc := base64.NewEncoder(base64.StdEncoding, &buf)
+	enc.Write([]byte{0})
+	enc.Write([]byte(c.Mailer.Username))
+	enc.Write([]byte{0})
+	smtpPassword, err := c.Mailer.PasswordConfig.Pass()
+	if err != nil {
+		return err
+	}
+	enc.Write([]byte(smtpPassword))
+	enc.Close()
+	authLine = fmt.Sprintf("AUTH PLAIN %s", buf.String())
+	return nil
+}
+
+func main() {
+	app := cmd.NewAppShell("mailcatch", "A simple SMTP server")
+	app.Action = func(c cmd.Config, stats statsd.Statter, auditlogger *blog.AuditLogger) {
+		err := calculateAuth(c)
```

Comment:
Since this is a test, rather than having a function to calculate auth generically from a Config, let's simplify by hardcoding the auth (as we do in mail/mailer_test.go).